### ISSUE TITLE
Vita3K on Imac

### ISSUE THREAD

**Issue by niekvlessert** [Vita3K/Vita3K#3119]
Hi,

I started the good old Shovel Knight on vita3k. Runs fine on Windows, however on my Intel imac with Ventura 13.6.1 I get this error, probably has something to do with Metal/Vulkan compatability.

[23:21:29.253] |W| [operator()]: Stubbed sceNetCtlInit import called. (Stub)
[23:21:29.253] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x3c733316 sceSslInit at 0x803081d0
[23:21:29.257] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x214926d9 sceHttpInit at 0x803081e0
[23:21:29.261] |I| [export_sceSysmoduleLoadModule]: Loading module ID: SCE_SYSMODULE_NP_TUS
[23:21:29.261] |W| [operator()]: Unimplemented sceNpTusInit import called.
[23:21:29.268] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x1cd8f88e fmodf at 0x80309bb0
[mvk-error] VK_ERROR_FORMAT_NOT_SUPPORTED: vkCreateRenderPass(): Attachment format VK_FORMAT_R5G6B5_UNORM_PACK16 is not supported on this device.
[mvk-error] VK_ERROR_FORMAT_NOT_SUPPORTED: VkFormat VK_FORMAT_R5G6B5_UNORM_PACK16 is not supported on this device.
[mvk-error] VK_ERROR_FORMAT_NOT_SUPPORTED: vkCreateRenderPass(): Attachment format VK_FORMAT_R5G6B5_UNORM_PACK16 on this device does not support the VkFormat attachment capabilities required by the subpass at index 0.
[23:21:29.290] |C| [operator()]: Unhandled C++ exception. vk::Device::createRenderPass: ErrorFormatNotSupported
libc++abi: terminating due to uncaught exception of type vk::FormatNotSupportedError: vk::Device::createRenderPass: ErrorFormatNotSupported
zsh: abort      /Applications/Vita3K.app/Contents/MacOS/Vita3K

**Comment by qeeg** [comment]
Dude, why are you playing Shovel Knight on an emulator for any reason other than testing it? It's been on PCs since it launched lol

**Comment by niekvlessert** [comment]
Cause I can of course. :)

**Comment by qeeg** [comment]
There is literally no point to it though, and doing so when you could literally just run the PC version (which is great in this case) is fucking stupid

**Comment by niekvlessert** [comment]
Well I'd say this is not the only game having this issue, it's simply a bug in the (Intel?) mac version of Vita3K, so worth fixing I think.

**Comment by qeeg** [comment]
Nah, M*c users don't deserve software, you're supporting one of the worst companies in the tech space right now.

**Comment by Macdu** [comment]
Just because you can is a sufficient reason.  
@qeeg what you are saying does not help in any way and is not nice, please refrain from making such comments.
This error can be fixed. But because it only happens on mac computers with iGPUs and is a bit annoying to fix, it has low priority right now sorry.

**Comment by niekvlessert** [comment]
Ok, thanks for the information. Not sure what you mean by iGPU though, usually one means a GPU in the CPU, this one has a Radeon Pro 570 4 GB though. I'll have a look myself maybe, if compiling everything is easy enough, maybe some hack is possible like they did here. https://github.com/cemu-project/Cemu/issues/290

**Comment by Macdu** [comment]
Thank you for the link, looks like the r5g6b5 format support is only enabled on macos 12 and higher (even though the apple website reports that it is available on macos 11, weird). For compatibility, Vita3K ships with a macos 11 compatible dylib.

If you have macos 12 or more, could you install moltenvk (brew install molten-vk), delete the libMoltenVK.dylib in Vita3K folder and replace it by the one installed with molten-vk?

**Comment by Macdu** [comment]
@qeeg Another comment like this and you get blocked from Vita3K repositories

**Comment by niekvlessert** [comment]
That's a nice trick to see what happens!

But if a lot of games use this r5g6b5, and with my limited experience I already found 2, what good will this compatibility with older MacOS versions do?

**Comment by niekvlessert** [comment]
Nope, some other dependency probably:

[23:28:19.715] |I| [load_module]: Loading module "vs0:sys/external/libssl.suprx"
[23:28:19.721] |I| [load_self]: Loaded module segment 0 @ [0x8105A000 - 0x810C0EA4 / 0x81000000] (size: 0x00066EA4) of module vs0:sys/external/libssl.suprx
[23:28:19.721] |I| [load_self]: Loaded module segment 1 @ [0x807F4000 - 0x807F9188 / 0x81067000] (size: 0x00005188) of module vs0:sys/external/libssl.suprx
[23:28:19.721] |I| [load_self]: Linking SELF vs0:sys/external/libssl.suprx...
[23:28:19.721] |D| [load_var_exports]: 	NID 0x936c8a78 (SYSLYB) at 0x810ac9b0
[23:28:19.721] |D| [load_var_exports]: 	NID 0x6c2224ba (NID_MODULE_INFO) at 0x810a6c70
[23:28:19.721] |I| [load_module]: Module SceLibSsl (at "vs0:sys/external/libssl.suprx") loaded
[23:28:19.721] |W| [operator()]: Stubbed sceNetCtlInit import called. (Stub)
[23:28:19.721] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x3c733316 sceSslInit at 0x803081d0
[23:28:19.723] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x214926d9 sceHttpInit at 0x803081e0
[23:28:19.725] |I| [export_sceSysmoduleLoadModule]: Loading module ID: SCE_SYSMODULE_NP_TUS
[23:28:19.725] |W| [operator()]: Unimplemented sceNpTusInit import called.
[23:28:19.730] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x1cd8f88e fmodf at 0x80309bb0
[23:28:19.754] |C| [operator()]: Unhandled C++ exception. vk::Device::createRenderPass: ErrorFormatNotSupported
libc++abi: terminating due to uncaught exception of type vk::FormatNotSupportedError: vk::Device::createRenderPass: ErrorFormatNotSupported

**Comment by niekvlessert** [comment]
I tried to fork and build without the --bottle-tag=montery, but it has been a while since I used actions... ;)

     - name: Set up build environment (macos-latest)
        run: |
          brew install ccache ninja create-dmg
          brew fetch --force --bottle-tag=monterey boost openssl@3 molten-vk
          brew install $(brew --cache --bottle-tag=monterey boost)
          brew install $(brew --cache --bottle-tag=monterey molten-vk)
          brew reinstall $(brew --cache --bottle-tag=monterey openssl@3)
          echo "/usr/local/opt/ccache/libexec" >> $GITHUB_PATH
          ccache --set-config=compiler_check=content
        if: matrix.os == 'macos-latest'

**Comment by niekvlessert** [comment]
Managed to build them, but still:

[00:30:23.929] |D| [load_var_exports]: 	NID 0x6c2224ba (NID_MODULE_INFO) at 0x810a6c70
[00:30:23.929] |I| [load_module]: Module SceLibSsl (at "vs0:sys/external/libssl.suprx") loaded
[00:30:23.929] |W| [operator()]: Stubbed sceNetCtlInit import called. (Stub)
[00:30:23.929] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x3c733316 sceSslInit at 0x803081d0
[00:30:23.932] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x214926d9 sceHttpInit at 0x803081e0
[00:30:23.934] |I| [export_sceSysmoduleLoadModule]: Loading module ID: SCE_SYSMODULE_NP_TUS
[00:30:23.934] |W| [operator()]: Unimplemented sceNpTusInit import called.
[00:30:23.940] |T| [log_import_call]: [LLE] TID: 6   FUNC: 0x1cd8f88e fmodf at 0x80309bb0
[00:30:23.965] |C| [operator()]: Unhandled C++ exception. vk::Device::createRenderPass: ErrorFormatNotSupported
libc++abi: terminating due to uncaught exception of type vk::FormatNotSupportedError: vk::Device::createRenderPass: ErrorFormatNotSupported

**Comment by Macdu** [comment]
@niekvlessert If you are interested, using a similar approach to RPCS3, you can go to 
https://github.com/Vita3K/Vita3K/blob/62a72bd7c6c73370b948cace697edfe17f39365f/vita3k/renderer/src/vulkan/gxm_to_vulkan.cpp#L451
and replace vk::Format::eR5G6B5UnormPack16 by vk::Format::eR8G8B8A8Unorm

**Comment by niekvlessert** [comment]
Had to change another one ~760.
